src/main/java/com/google/sps/servlets/GmailServlet.java

```diff
@@ -66,15 +71,58 @@ public void doGet(
     assert googleCredential != null
         : "Null credentials (i.e. unauthenticated requests) should already be handled";
 
-    // Get messageIds from Gmail
     GmailClient gmailClient = gmailClientFactory.getGmailClient(googleCredential);
-    List<Message> messages = gmailClient.listUserMessages("");
 
-    // convert messageIds to JSON object and print to response
-    Gson gson = new Gson();
-    String messageJson = gson.toJson(messages);
+    int nDays;
+    int mHours;
+    try {
+      nDays = Integer.parseInt(request.getParameter("nDays"));
+      mHours = Integer.parseInt(request.getParameter("mHours"));
+    } catch (NumberFormatException e) {
+      response.sendError(400, "nDays and mHours must be integers");
+      return;
+    }
 
-    response.setContentType("application/json");
-    response.getWriter().println(messageJson);
+    if (nDays < 0 || mHours < 0) {
+      response.sendError(400, "nDays and mHours must be positive");
+      return;
+    }
+
+    if (mHours > nDays * 24) {
+      response.sendError(400, "mHours must represent less time than nDays");
+      return;
+    }
+
+    GmailResponse gmailResponse = generateGmailResponse(nDays, mHours, gmailClient);
+    JsonUtility.sendJson(response, gmailResponse);
+  }
+
+  /**
+   * Creates a GmailResponse object
+   *
+   * @param nDays number of days unread emails should be checked for
+   * @param mHours unread emails from the last mHours hours should be used to calculate the
+   *     "unreadEmailsFromMHours" statistic
+   * @param gmailClient GmailClient implementation with valid credential
+   * @return GmailResponse object
+   * @throws IOException if an issue occurs with the Gmail Service while obtaining emails
+   */
+  private GmailResponse generateGmailResponse(int nDays, int mHours, GmailClient gmailClient)
+      throws IOException {
+    GmailClient.MessageFormat messageFormat = GmailClient.MessageFormat.METADATA;
+
+    List<Message> unreadMessages =
+        GmailResponseUtility.getUnreadEmailsFromNDays(gmailClient, messageFormat, nDays);
+    int unreadEmailsFromNDays = unreadMessages.size();
+    int unreadEmailsFromMHours = GmailResponseUtility.countEmailsFromMHours(unreadMessages, mHours);
+    int unreadImportantEmailsFromNDays = GmailResponseUtility.countImportantEmails(unreadMessages);
+    String mostFrequentSender =
+        GmailResponseUtility.findMostFrequentSender(unreadMessages).orElse("");
```

Comment:
Can you use dependency injection to access `GmailResponseUtility`? That way you will be able to test the logic of `GmailServlet` separately (which will also make testing it very simple)

Reply:
I am not actually convinced this simplifies things (without sacrificing test coverage), but I look forward to the result of the refactoring.

Reply:
Done